gate-core/src/main/java/com/netflix/spinnaker/gate/services/DefaultProviderLookupService.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2016 Netflix, Inc.
+ * Copyright 2023 Apple, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.gate.services;
+
+import com.netflix.spinnaker.fiat.model.Authorization;
+import com.netflix.spinnaker.gate.services.internal.ClouddriverService;
+import com.netflix.spinnaker.security.AuthenticatedRequest;
+import java.util.Collection;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+import org.springframework.util.CollectionUtils;
+
+@Log4j2
+@Service("providerLookupService")
+@RequiredArgsConstructor
+public class DefaultProviderLookupService implements ProviderLookupService, AccountLookupService {
+  private static final String FALLBACK = "unknown";
+  private final ClouddriverService clouddriverService;
+
+  private volatile List<ClouddriverService.AccountDetails> accountsCache = List.of();
```

Comment:
why volatile instead of AtomicReference?

Reply:
Simpler way to represent the same exact use case. `AtomicReference` is mostly useful when you want to use CAS operations. For doing barrier-enforced reads and writes to values, a `volatile` field is sufficient (and can be further optimized to use either a `VarHandle` or back to `AtomicReference` when using `getAcquire` and `setOpaque` operations, but this is mostly relevant to non-x86 architectures with more interesting memory models).